README.md
# IIS Express for VSCode
This extension gives you the power to run a folder open in Visual Studio Code as a website using an IIS Express web server.

[![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/warren-buckley.iis-express?logo=visual-studio-code&style=for-the-badge)](https://marketplace.visualstudio.com/items?itemName=warren-buckley.iis-express)
[![Visual Studio Marketplace Installs](https://img.shields.io/visual-studio-marketplace/i/warren-buckley.iis-express?logo=visual-studio-code&style=for-the-badge)](https://marketplace.visualstudio.com/items?itemName=warren-buckley.iis-express)
[![Visual Studio Marketplace Downloads](https://img.shields.io/visual-studio-marketplace/d/warren-buckley.iis-express?logo=visual-studio-code&style=for-the-badge)](https://marketplace.visualstudio.com/items?itemName=warren-buckley.iis-express)
[![Visual Studio Marketplace Rating](https://img.shields.io/visual-studio-marketplace/r/warren-buckley.iis-express?logo=visual-studio-code&style=for-the-badge)](https://marketplace.visualstudio.com/items?itemName=warren-buckley.iis-express)

## Sponsorware
<a href="https://github.com/sponsors/warrenbuckley"><img src="https://github.githubassets.com/images/modules/site/sponsors/pixel-mona-heart.gif" align="left" height="45" /></a>
This is a **free extension** I have made for VSCode, if you find it useful to yourself or your business then <a href="https://github.com/sponsors/warrenbuckley">I would love you to consider sponsoring me on Github</a> please

## Features
* Auto opens folder in browser
* Super simple way from the command pallete to start & stop the website
* See ouput from the IIS Express command line directly in Visual Studio
* Support for PHP via `ApplicationHost.config` common changes for all sites running with IIS Express

### Requirements
* Windows Machine (Sorry not for Linux, OSX or Web)

## How do I Install IIS Express for VSCode?
Open the command pallete in VSCode & type **ext install** then search for **IIS Express** alternatively you can [download the IIS Express for VSCode extension on the marketplace](https://marketplace.visualstudio.com/items?itemName=warren-buckley.iis-express)

![Installing IIS Express for VSCode](images/iis-express-install.gif)

## Settings for IIS Express
IIS Express for VSCode automatically generates a default config file for you & stores it at `.vscode/iisexpress.json`
You can modify this file on disk inside VSCode to get intellisens/autocompletion for the available properties for you to set

Setting | Alias | Description|Required
--------|-------|------------|--------
Port Number|port|The port number you wish to use for IIS Express|✔
Folder Path of Site|path|This property is optional & allows you to set an absolute path to the folder or subfolder as the root of your site/project for IIS Express to run. Additionally this support relative paths to the workspace folder such as `./my-sub-folder`|❌
CLR|clr|This property is optional & allows you to set AppPool CLR version, 2.0 vs 4.0|❌
Protocol|protocol|This property is optional & allows you to set http or https|❌
Url Suffix|url|This property is optional & allows you to set the page URL you wish to open eg: '/about/the-team'|❌

### Default Settings
By default with the file not generated it will create the following file for you on disk and setting a random port number.
```json
{
  "port": 18420,
  "path": "./",
  "clr": "v4.0",
  "protocol": "http"
}
```

## Global Settings for IIS Express for VSCode
Below is a table of settings that can be used to configure IIS Express for VSCode. You can easily get to these settings by using the settings (cog) icon in the IIS Express pane found by default under the file explorer.

Setting | Alias | Description | Default Value
--------|-------|-------------|------
Path to IISExpress.exe|iisexpress.iisExpressPath|An absolute path to **IISExpress.exe** such as `C:\Program Files\IIS Express\iisexpress.exe`|null - We try to use default program path location
Path to AppCmd.exe|iisexpress.appcmdPath|An absolute path to **appcmd.exe** such as `C:\Program Files\IIS Express\appcmd.exe`|null - We try to use default program path location
Auto launch browser|iisexpress.autoLaunchBrowser|An option to disable or enable the browser from launching the site when IIS Express starts. By default this is **true**|true
Open in browser|iisexpress.autoLaunchBrowser|Decide which browser to auto launch the site with when `iisexpress.autoLaunchBrowser` is set to **true**|default (default, chrome, msedge, firefox, opera)

## Available commands
* **Start Website** - Start the current folder run as a website
* **Stop Website** - Stop the current folder run as a website
* **Restart Website** - Restart (Stop & Start) the current folder run as a website
* **Become a supporter** - Opens a webpage to my GitHub Sponsor page if you wish to donate & become a sponsor
* **Open Settings** - This directly opens VSCode settings for IIS Express

## How do I use the extension?
The VSCode command pallete is your friend. Press **F1 on the keyboard and start typing IISExpress** to see the list of commands to use.
Alternatively is version 1.2.0 and newer there is a new panel view that by default lives under the list of files in the file explorer, in newer versions of VSCode you are able to drag and drop this pane around to a new location if you prefer to put it elsewhere.

## How do I use HTTPS for my site?
To enable HTTPS for your site use the settings file `.vscode/iisexpress.json` set the property **protocol** to **https** and ensure the port property is in the range between `44300 - 44399`

## How do I change keymappings for IIS Express for VSCode
By default IIS Express for VSCode binds to the following keyboard shortcuts, you may prefer to change these keybindings if you find them conflicting in your workflows.
The best way to do this is to [read the official documentation from VSCode](https://code.visualstudio.com/docs/getstarted/keybindings).

Keyboard Shortcut|Command
-----------------|-------
ctrl+f5| Start Site
shift+f5| Stop Site
ctrl+shift+f5| Restart Site



CHANGELOG.md
# Changelog
## Version 1.2.0
### 🚀 Features

- GitHub Sponsors & Friendly Sponsorware Message @warrenbuckley (#159)
- Added Telemetry @warrenbuckley (#125)
- Support for relative paths for the site folder to run in iisexpress.json config file @warrenbuckley (#118)
- Custom IIS Express View in Explorer for easier access to commands @warrenbuckley (#113)
- Add in VSCode custom when context for enabling & disabling commands @warrenbuckley (#109)
- VSLiveShare - Auto shares IIS Express as a shared server in a LiveShare session @warrenbuckley (#107)
- Adds new config option iisexpress.openInBrowser to specify which browser to use with auto launch @warrenbuckley (#100)
- Adds new config option to toggle auto launch of browser @warrenbuckley (#87)
- EXE paths to IISExpress & AppCmd can be configured @warrenbuckley (#86)

### 🐛 Bug Fixes

- Update JSON Schema to use local file @warrenbuckley (#97)

### 🧰 Maintenance

- Lots of node dependencies updated

## Version 1.1.2
* **Bug Fix:** Version 1.1.1 fixed starting & stopping sites issue & having to cold reboot VSCode however the restart was overlooked. Thanks to @swensorm for the PR & fix for this :)

## Version 1.1.1
* **Bug Fix:** Fix issue where the site could not be restarted and required users to close VSCode to start the site up again

## Version 1.1.0
* **New:** Adds a new feature where IIS Express can be auto installed - by downloading the MSI from Microsoft & installing it for you when IIS is not detected

## Version 1.0.0
* **New:** `.vscode/iisexpress.json` now supports a new property `protocol` which is an optional property & allows you to set the protocol to either `http` or `https` By default this is set to `http` If you do use `https` then IIS Express expects the port number to be within a range of 44300 - 44399
* **New:** Uses `ApplicationHost.config` allowing for common changes for all sites running IIS Express. This allows support for using PHP for example.

### PHP Support
If you wish to use PHP then you will need to install PHP for Windows and then use the following commands in the console

**NOTE:** This alternatively may be `\Program Files (x86)\IIS Express\`
```
cd "c:\Program files\IIS Express\"
```

**NOTE:** The path to `php-cgi.exe` may be different for you so update the two commands below.

```
appcmd set config /section:system.webServer/fastCGI /+[fullPath='"C:\Program Files (x86)\P  HP\php-cgi.exe"']

appcmd set config /section:system.webServer/handlers /+[name='PHP_via_FastCGI',path='*.php',verb='*',modules='FastCgiModule',scriptProcessor='"C:\Program Files (x86)\PHP\php-cgi.exe"',resourceType='Unspecified']

appcmd set config /section:system.webServer/defaultDocument /+"files.[value='index.php']"
```

## Version 0.0.7
* **New:** `.vscode/iisexpress.json` now supports a new property `url` which is an optional property & allows you to set the URL you wish to open eg: '/about/the-team'
* **New:** `.vscode/iisexpress.json` now supports a new property `clr` which is an optional property allowing you to set the CLR to run your .NET website (Thanks to @avieru)

## Version 0.0.6
* **New:** Restart site option
* **New:** Keyboard shortcuts to start (Ctrl+F5), stop (Shift+F5) & restart site (Ctrl+Shift+F5)
* **New:** Clicking the status bar icon now opens the site in your browser using the 'start' command & you no longer needn to copy & paste the URL from the ouput window anymore
* **New:** `.vscode/iisexpress.json` now supports a new property `path` which allows you to set a subfolder inside your folder as the path of your site to run. This must be a full path & not relative
* **New:** JSON Schema validation & auto-completion of the `.vscode/iisexpress.json` file
* **Bug Fix:** Fixed logic that will now auto create the `.vscode/iisexpress.json` config file with the random assigned port if it does not already exist
* **Removes** The old UI of launching the extension/command & using a quick pick options to start & stop the site, as we could not assign a keyboard shortcut to the quick pick items


Adds the following:
* Seperate commands for Start, Stop & New Restart option
* With new commands we can assign key bindings aka keyboard shortcuts same as the main VS
* The status bar when clicked now opens the site in your browser using the 'start' command & don't want to paste & copy the URL from the ouput window anyymore
* JSON Schema for the VS Code IISExpress config. Ensure port is specified config value & within correct range of valid port numbers according to IIS Express

Removes the following
* The old UI of launching the extension/command & using a quick pick options to start & stop the site, as we could not assign a keyboard shortcut to the quick pick items

## Version 0.0.5
* **Bug Fix:** Fixes problem with extension from loading correctly after updating to VSCode 1.3

## Version 0.0.4
* **New:** Community PR from @czhj that supports Chinese characters in output build log window

## Version 0.0.3
* **New:** Port number setting can be set in .vscode/iisexpress.json if you wish to override random port number

You need the following file added `.vscode/iisexpress.json` and then you can set the port setting like so

```
{
    "port":8081
}
```

## Version 0.0.2
* **New:** Port number is now random

## Version 0.0.1
* Initial release


src/IISExpress.ts
import * as vscode from 'vscode';
import * as process from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as settings from './settings';
import * as telemtry from './telemetry';

// External libraries
import { v4 as uuidv4 } from 'uuid';
import * as iconv from 'iconv-lite';
import TelemetryReporter from 'vscode-extension-telemetry';


export interface IExpressArguments {
	path: string;
	port: number;
	clr: settings.clrVersion;
	protocol: settings.protocolType;
}

export class IISExpress {
	private _iisProcess!: process.ChildProcessWithoutNullStreams;
	private _iisPath: string;
	private _iisAppCmdPath: string;
	private _args!: IExpressArguments;
	private _output!: vscode.OutputChannel | null;
	private _statusbar!: vscode.StatusBarItem;
	private _statusMessage!: string;
	private _context: vscode.ExtensionContext;
	private _reporter: TelemetryReporter;

	constructor(iisPath: string, appCmdPath: string, context:vscode.ExtensionContext, reporter:TelemetryReporter){
		this._iisPath = iisPath;
		this._iisAppCmdPath = appCmdPath;
		this._context = context;
		this._reporter = reporter;
	}

	public startWebsite(options: settings.Isettings) {

		// Verify process not already running, so if we have a PID (process ID) it's running
		if(this._iisProcess !== undefined && this._iisProcess.killed === false){
			// Display error message that it's already running
			vscode.window.showErrorMessage('IIS Express is already running',{modal: true});

			// Stop the method/function from running
			return;
		}

		if(!options){
			vscode.window.showErrorMessage('No options found in .vscode/iisexpress.json',{modal: true});

			// Stop the method/function from running
			return;
		}

		this._args = {
			// Get IIS Port Number from config file
			port: options.port,

			// Folder to run as the arg
			path: options.path ? options.path : <string>vscode.workspace.rootPath,

			// CLR version, yes there are still people on 3.5 & default back to v4 if not set
			clr: options.clr ? options.clr : settings.clrVersion.v40,

			// If no protocol set fallback to http as opposed to https
			protocol: options.protocol ? options.protocol : settings.protocolType.http
		};



		// The path stored in options could be a relative path such as './' or './child-sub-folder'
		// It may also include the legacy full path of the workspace rootpath
		// Which caused issues - when checked into source control & other users did not have same folder structure
		// or if you moved the folder on your own machine - it would no longer map correctly
		const resolvedPath = path.resolve(<string>vscode.workspace.rootPath, this._args.path);
		this._args.path = resolvedPath;

		// Verify folder exists on disk (in case relative path used & selected wrong thing)
		if(fs.existsSync(resolvedPath) === false){
			vscode.window.showErrorMessage(`The folder does not exist ${resolvedPath} & IIS Express can not run.`, {modal: true});

			// Stop the method/function from running
			return;
		}

		// Create output channel & show it
		this._output = this._output || vscode.window.createOutputChannel('IIS Express');
		this._output.show(vscode.ViewColumn.Three);

		// Site name is the name of the workspace folder & GUID/UUID
		// Need to append a UUID as could have two folders/sites with same name
		const siteName = path.basename(vscode.workspace.rootPath as string) + "-" + uuidv4();

		// If user is using HTTPS & port not in range of auto-approved port numbers (44300-44399)
		// Then display an error & stop process
		if(this._args.protocol === settings.protocolType.https && (this._args.port >= 44300 && this._args.port <=44399) === false){
			// Using HTTPS but not using a port within the range that supports SSL
			vscode.window.showErrorMessage('When using HTTPS you need to use ports 44300 - 44399 in .vscode/iisexpress.json',{modal: true});

			// Stop the method/function from running
			return;
		}

		// Add the site to the config (which will invoke/run from iisexpress cmd line)
		// Not done as async - so we wait until this command completes
		try {
			process.execFileSync(this._iisAppCmdPath, ['add', 'site', `-name:${siteName}`, `-bindings:${this._args.protocol}://localhost:${this._args.port}`, `-physicalPath:${this._args.path}`]);
		} catch (error) {
			console.log(error);
			this._reporter.sendTelemetryException(error, {"appCmdPath": this._iisAppCmdPath, "appCmd": `add site -name:${siteName} -bindings:${this._args.protocol}://localhost:${this._args.port} -physicalPath:${this._args.path}`});
		}

		// Based on the CLR chosen use the correct built in AppPools shipping with IISExpress
		const appPool = this._args.clr === settings.clrVersion.v40 ? "Clr4IntegratedAppPool" : "Clr2IntegratedAppPool";

		// Assign the apppool to the site
		// appcmd set app /app.name:Site-Staging-201ec232-2906-4052-a431-727ec57b5b2e/ /applicationPool:Clr2IntegratedAppPool
		try {
			process.execFileSync(this._iisAppCmdPath, ['set', 'app', `/app.name:${siteName}/`, `/applicationPool:${appPool}`]);
		} catch (error) {
			console.log(error);
			this._reporter.sendTelemetryException(error, {"appCmdPath": this._iisAppCmdPath});
		}

		// Log telemtry
		telemtry.updateCountAndReport(this._context, this._reporter, telemtry.keys.start);
		telemtry.updateCountAndReport(this._context, this._reporter, telemtry.keys.sponsorware);

		// This is the magic that runs the IISExpress cmd from the appcmd config list
		this._iisProcess = process.spawn(this._iisPath, [`-site:${siteName}`]);

		// Create Statusbar item & show it
		this._statusbar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);

		// Set props on statusbar & show it
		this._statusbar.text = `$(browser) ${this._args.protocol}://localhost:${this._args.port}`;
		this._statusMessage = `Running folder '${this._args.path}' as a website on ${this._args.protocol}://localhost:${this._args.port} on CLR: ${this._args.clr}`;
		this._statusbar.tooltip = this._statusMessage;
		this._statusbar.command = 'extension.iis-express.open';
		this._statusbar.show();

		// Open browser
		const autoLaunchBrowser:boolean = vscode.workspace.getConfiguration().get<boolean>('iisexpress.autoLaunchBrowser', true);

		if(autoLaunchBrowser){
			this.openWebsite(options);
		}

		// Used to enable/disable commands & to know when a site is running
		vscode.commands.executeCommand('setContext', 'iisexpress:siterunning', true);

		// Attach all the events & functions to iisProcess
		this._iisProcess.stdout.on('data', (data: string) =>{
			data = this.decode2gbk(data);
			this._output!.appendLine(data);
		});

		this._iisProcess.stderr.on('data', (data: string) => {
			data = this.decode2gbk(data);
			this._output!.appendLine(`stderr: ${data}`);
		});

		this._iisProcess.on('error', (err:Error) => {
			const message = this.decode2gbk(err.message);
			this._output!.appendLine(`ERROR: ${message}`);
		});

		this._iisProcess.on('close', () =>{

			// Tidying up - so we remove the entry from appcmd
			// As we use a uuid every time we start a site we need to do housekeeping & clean up
			// When we are finished - this happens for:

			//* Will happen when users stops with CTRL+F5
			//* Close from the systray icon
			//* Restart of site

			// Delete any existing entries for the site using appcmd
			// Not done as async - so we wait until this command completes
			try {
				process.execFileSync(this._iisAppCmdPath, ['delete', 'site', `${siteName}`]);
			} catch (error) {
				console.log(error);
				this._reporter.sendTelemetryException(error, {"appCmdPath": this._iisAppCmdPath, "appCmd": `delete site ${siteName}`});
			}
		});
	}

	public stopWebsite(){
		// Ensure process is not null & is running
		if(this._iisProcess === undefined || this._iisProcess.killed === true){
			// Display error message
			vscode.window.showErrorMessage('IIS Express is not running', {modal:true});

			// Stop the method/function from running
			return;
		}

		// Kill the process - which will also hook into the exit event to remove the config entry
		this._iisProcess.kill('SIGINT');

		// Used to enable/disable commands & to know when a site is running
		vscode.commands.executeCommand('setContext', 'iisexpress:siterunning', false);

		// Log telemtry
		telemtry.updateCountAndReport(this._context, this._reporter, telemtry.keys.stop);

		// Clear the output log
		this._output!.clear();
        this._output!.hide();
        this._output!.dispose();
		this._output = null;

		// Remove the statusbar item
		this._statusbar.hide();
		this._statusbar.dispose();

	}

	public openWebsite(options: settings.Isettings){

		let browserConfig = vscode.workspace.getConfiguration().get<string>("iisexpress.openInBrowser", "default");
		let browser:string = browserConfig.toLocaleLowerCase() === "default" ? "" : browserConfig;

		if (options && options.url) {
			// We have a starting URL set - but lets ensure we strip starting / if present
			let startUrl = options.url.startsWith('/') ? options.url.substring(1) : options.url;

			// Start browser with start url
			process.exec(`start ${browser} ${this._args.protocol}://localhost:${this._args.port}/${startUrl}`);
    	} else {
			// Uses the 'start' command & url to open default browser
			process.exec(`start ${browser} ${this._args.protocol}://localhost:${this._args.port}`);
		}
	}

	public restartSite(options : settings.Isettings){
		// If we do not have an iisProcess/website running
		if(!this._iisProcess){
			// Then just do a start site...
			this.startWebsite(options);
		}
		else {
			// It's already running so stop it first then, start it
			this.stopWebsite();
			this.startWebsite(options);
		}

		// Log telemtry
		telemtry.updateCountAndReport(this._context, this._reporter, telemtry.keys.restart);
	}

    private decode2gbk(data: string): string {
		const buffer = Buffer.from(data);
 		return iconv.decode(buffer, 'gbk');
	}

}
